### ISSUE TITLE
production readiness

### ISSUE THREAD

**Issue by mfulton26** [GoogleChromeLabs/sw-appcache-behavior#7]
If I understand correctly, Chrome will remove AppCache support on/around Apr 28 (https://www.chromestatus.com/feature/6192449487634432, https://www.chromestatus.com/features/schedule).

My team uses AppCache in production for reasons explained [here](https://bugs.chromium.org/p/chromium/issues/detail?id=582750#c32):

> WKWebView does not support service workers but it does support AppCache. Chrome on Android also supports AppCache (currently). If AppCache support is removed in Chrome then the suggested alternative, service workers, is not viable for a cross-platform web app that runs inside of a WebView (e.g. via react-native-web-view on iOS and Android).

The README for these packages currently state:

> This project is still a work in progress. Please experiment with it on a non-production version of your site, and open issues with feedback or bug reports if you run in to problems.

Is there a goal for this AppCache polyfill to be production ready before AppCache is removed from Chrome?

**Comment by jeffposnick** [comment]
While I believe that this polyfill is functional and complies with the AppCache specification, that language is in place because we have not heard many examples of this polyfill being used "in the wild".

I'd very much appreciate feedback based on your real-world usage. But the (perhaps obvious) recommendation is to try this polyfill out on a local copy of your project first to confirm that there are no edge cases, before deploying it widely to production.

**Comment by mfulton26** [comment]
Is there a way to test on an Android emulator/phone with an Android WebView that does not have AppCache in it? I've never tried getting a dev channel Android WebView before…

**Comment by jeffposnick** [comment]
In Chrome, once there's a service worker registered and a page is loaded in scope of that service worker, AppCache is disabled. (I.e. the service worker always "wins", even if there's still a `manifest` attribute on the `<html>`.)

My assumption is that would also apply to Android WebView, but it's not something I've personally confirmed. If you're seeing otherwise, we can escalate that as a bug with WebView.

**Comment by mfulton26** [comment]
I see. Thank you for the reply! I am out on vacation this week but next week when I'm back at work I will see if we can test this out, etc. Thank you for building this! It looks very promising to satisfy our current use case. We can leave this issue open if you'd like and until we report back our findings or close this and if we find issues we can open new ones. Whichever you'd prefer. Thank you again.

**Comment by jeffposnick** [comment]
I'd love to leave this open and track what I hope will be a positive affirmation that the polyfill will meet your needs.

**Comment by mfulton26** [comment]
the app I need to get this working in is a react-app, I'm not sure yet how to get webpack to expose the `mjs` and `umd.js` files yet… any plans on making any demo/example apps? one for `create-react-app` would be awesome

**Comment by jeffposnick** [comment]
You can add

```
<script type="module">
  import {init} from '/path/to/appcache-polyfill-window/build/index.mjs';
  init().then(() => navigator.serviceWorker.register('sw.js'));
</script>
```

to the HTML that's output by React. The exact path to use would depend on where you put your `appcache-polyfill-window`. You could, for instance, use the [public folder](https://create-react-app.dev/docs/using-the-public-folder/) in `create-react-app` and store your `index.mjs` there.

Alternatively, if you would rather bundle `appcache-polyfill-window` along with your main `webpack` output that `create-react-app` manages, you should be able to

```
npm install --save-dev appcache-polyfill-window
```

and then in your `index.js` (or some other JS file):

```
import {init} from 'appcache-polyfill-window';

init();
```

**Comment by mfulton26** [comment]
I am trying this out. One thing that I still have a question about is the `importScripts` statement in `sw.js`… I need a path to the `appcache-polyfill-sw/index.umd.js` which I either need to manually copy it there (not ideal), write a `prebuild` script to copy it there (better but still not ideal), or find a way to tell webpack to put it somewhere… not sure yet. Thank you for your help.

FYI: I did copy these polyfill files manually to my `public` folder so that I could test them out and things seem to be working so far. We have several more things to test so I will keep you posted.

On a related note, should I expect to see any logs from these polyfill scripts or is silence a good sign? 😄

**Comment by mfulton26** [comment]
ah, I think I can use `copy-webpack-plugin` to copy from `node_modules` to `build`…

**Comment by mfulton26** [comment]
may have found an issue: we have an auto-reload feature in our app that invokes `applicationCache.update()` periodically; when it detects `applicationCache.UPDATEREADY` it calls `applicationCache.swapCache()` and then calls `location.reload()`. This works fine in `WKWebView` on iOS and works with Chrome's current native AppCache support.

 Might the cache swapping with this polyfill be asynchronous or something and we need to wait to reload?

**Comment by mfulton26** [comment]
After digging into this a bit more it looks like `appcache-polyfill-window` doesn't define or override `window.applicationCache`, `window.applicationCache.status`, `window.applicationCache.update()`, `window.applicationCache.swapCache()`, `window.applicationCache.UPDATEREADY`, `window.applicationCache.onupdateready`, `window.applicationCache.addEventListener`, etc.

The offline cache population, updating, and usage seems to work great from what I can tell. The only missing piece I'm aware of is defining/overriding the `window.applicationCache` object itself.

**Comment by jeffposnick** [comment]
Hello @mfulton26—you're correct in that this polyfill does not attempt to replicate the `window.applicationCache` interface, nor does it attempt to fire the associated events.

I've [updated the README](https://github.com/GoogleChromeLabs/sw-appcache-behavior/commit/ffd37ade55bb1bd2a6e2c733e9c6b603cf542c2b#diff-04c6e90faac2675aa89e2176d2eec7d8) to make that clear.

Some background on that decision: the idea behind this polyfill is to ease the transition from AppCache to service workers. Once using a service worker, a great deal of additional flexibility becomes possible. That flexibility means that, as you make changes to your service worker code to accomplish things like using [custom caching strategies](https://developers.google.com/web/fundamentals/instant-and-offline/offline-cookbook), you lose the ability to map what your arbitrary service worker code does into the old model of AppCache events/methods.

In other words, replicating the AppCache caching/serving experience inside of a service worker is meant as a starting point, allowing you to then customize things further and get a richer offline experience.

With that in mind, what we might be able to do is publish some "recipes" in the docs for this repo, along the lines of "if you used to do _xyz_ using the `window. applicationCache` interface, then something (roughly) along these lines inside of your service worker would be equivalent, assuming you are mostly just using the polyfill code inside your service worker. Some of those recipes would likely need to be based on using `postMessage()` to communicate between the `window` and service worker.

It sounds like your auto-reload feature might be the first candidate for such a recipe.

**Comment by mfulton26** [comment]
@jeffposnick I think for our needs we just need a way to get a callback/event for when the cache was updated so that we can trigger a page reload

any ideas on how we might expose that?

thank you

**Comment by jeffposnick** [comment]
Okay, that's actually a little more straightforward.

The cache is populated in the `window` context, at https://github.com/GoogleChromeLabs/sw-appcache-behavior/blob/ffd37ade55bb1bd2a6e2c733e9c6b603cf542c2b/packages/appcache-polyfill-window/index.ts#L66-L69

Something I could do is modify the `init()` method in `appcache-polyfill-window` so that it accepts a callback function, like

```
<script type="module">
  import {init} from '/path/to/appcache-polyfill-window/build/index.mjs';

  function myCachePopulatedCallback(urls) {
    // Your logic goes here.
  }

  init({
    cachePopulatedCallback: myCachePopulatedCallback,
  }).then(() => navigator.serviceWorker.register('sw.js'));
</script>
```

And then `appcache-polyfill-window` would call your callback function whenever all of the `cache.put(url, response)` operations for a given manifest version have completed, passing in the array of `urls` that was just added to the cache.

Would that be enough for your use case? Please note that with this approach, your callback would be invoked unconditionally with all of the URLs that are added to the cache, without any additional context about whether it's an update to a previously cached entry or anything like that.

**Comment by mfulton26** [comment]
That sounds good. If I wanted to periodically invoke an "update" I would just call `init()`, is that right? If nothing gets updated then the callback would not be invoked but if there are updates then once they are all updated the callback would be invoked with a list of URLs that were updated.

**Comment by jeffposnick** [comment]
Although I didn't anticipate that folks would call `init()` multiple times within the lifetime of a page load, calling it again should kick off a manifest re-download. Then yes, it would trigger the proposed new callback function following a `cache.put()`.

I can put out a new release of this library with support for that callback over the next few days if you want to try it out.

**Comment by mfulton26** [comment]
That would be perfect. Thank you.

**Comment by mfulton26** [comment]
Any update on this? I've got a change mostly ready on my end to incorporate this polyfill but it breaks our auto-update feature. It'd be great to get a way to make that work and finish this up. Thanks!

**Comment by jeffposnick** [comment]
Apologies for the delay.

I've just published `v0.2.0` of the libraries to `npm`, which includes this functionality. There's some refactoring that I still want to do, along with some test cases, but I figured I shouldn't delay in releasing something you could test out and give feedback on.

Let me know whether it behaves as you'd expect:

```
<script type="module">
  import {init} from '/path/to/appcache-polyfill-window/build/index.mjs';

  function myCachePopulatedCallback(urls) {
    // Your logic goes here.
  }

  init({
    cachePopulatedCallback: myCachePopulatedCallback,
  }).then(() => navigator.serviceWorker.register('sw.js'));
</script>
```

**Comment by jeffposnick** [comment]
...just wanted to confirm (wasn't clear from the 👍 emoji) that `v0.2.0` provides the functionality that you're looking for, @mfulton26.

**Comment by mfulton26** [comment]
Right, we're still not using this in production completely but from my local testing I believe the answer is yes. Part of me wishes I could polyfill all the events on `ApplicationCache` and not just `update` but that's all I really truly need. I think we can probably close this issue and if we identify any issues going forward we can open separate issues then. Thank you so much for providing this polyfill and providing the support needed. This has been great!

**Comment by jeffposnick** [comment]
Glad to hear it!
